testsuite/test_twofactor.py

```diff
@@ -147,3 +148,51 @@ def test_auto_logout(client, mock_time, listen):
     mock_time(t0 + timedelta(hours=3, minutes=0, seconds=5))
     assert not _access_homepage(client)
     assert auto_logout == [{'username': 'john'}]
+
+def test_rate_limit(client, mock_time, listen):
+    rate_limit_exceeded = listen(signals.rate_limit_exceeded)
+    t0 = datetime(2016, 6, 13, 12, 0, 0, tzinfo=utc)
+    now = None
+
+    def set_time(t):
+        nonlocal now
+        now = t
+        mock_time(t)
+
+    def try_login(correct_otp, expect_limit, expect_success):
+        device.last_t = -1
+        device.save()
+        resp = client.post('/accounts/login/', {
+            'username': 'john',
+            'password': 'pw',
+            'otp_token': _totp(device, now) if correct_otp else '123456',
```

Comment:
What, and forego the chance to see it fail? It would be one hell of an error message :)